### PR TITLE
Make the editor canvas friendly towards colored backgrounds

### PR DESCRIPTION
**Pull request by jasmussen**
One long term vision for the editor, is to allow you to load the theme stylesheet into the editor canvas itself, so you have an almost if not completely WYSIWYG experience where the theme background and fonts apply to the editor.

In the interim, we should make it easier to load _editor styles_ into the canvas. This already works today, but the block UI is not really ready for it — white backgrounds are used here and there, and borders are colored light grays which quickly disappear on some colors.

This PR takes an initial stab at alleviating that, by moving to using various levels of opacities of the darkest gray we have, to emulate the range of opaque grays we have. As such, this PR includes a range of those opacities. For example, shown on a white background, $dark-opacity-500 should look the same as $dark-gray-500, but the latter should look good on colored backgrounds as well.

This is going to be a fair bit of work, and this is only the beginnings of it. It styles the borders, the outermost toolbar borders, the naked mover/settings/more icons, and the title field, as well as the appender. In doing this, it now means you can at least change the background color and still sort of use the editor.

Asking for thoughts now, before I do more work — what do you think of this approach?

If there's positive agreement on this direction, I will fix the drag and drop shadows/colors. Note that this PR makes the UI work on _light_ backgrounds. If you cross a certain threshold of luminosity into the dark background territory, we need to invert the colors. This PR already contains the opacities for that, but we'll need to detect the background color and apply a body class, something like `body.has-dark-background`, so that I can invert the UI for that. Would appreciate thoughts on this too.

Video: https://cloudup.com/cDX4dvMYFvs

### CODE REVIEW

**Comment by paulwilde** [comment]
Looks good from what I saw in the video, and I also had the same considerations with different opacities based on if the BG is light/dark. To be honest, I cannot think of a different approach than this  to add background support... so I think this is the best approach.

Something I've picked up upon from adding this on a prior site built in Gutenberg and watching the video you've attached is that the hover states of the buttons need a white background as otherwise the border looks kinda ugly.

**Comment by jasmussen** [comment]
Cool, thanks for the initial sanity check, Paul. I'll let this branch sit for a bit further, but then take a stab at making further improvements. 

> Something I've picked up upon from adding this on a prior site built in Gutenberg and watching the video you've attached is that the hover states of the buttons need a white background as otherwise the border looks kinda ugly.

Yep — that's a solid point, and something I'll add. 🤘

**Comment by afercia** [comment]
Hmm... giving themes the power to modify the UI could potentially break any colors contrast ratio for accessibility. I'd tend to think Gutenberg should preserve the accessibility of its own interface, maybe reserving some light background around its controls. 

Also, using opacity will make a bit difficult calculating the contrast ratio... as we won't be able to pick up the actual color from the CSS (or browser inspector), inteaad we'll have to use tools that pick up the colors rendered on screen  🙂

**Comment by karmatosed** [comment]
In theory this seems like a good idea and I absolutely agree with breaking it down. Before I review though, how does (if at all) fit in with: https://github.com/WordPress/gutenberg/issues/5360?

**Comment by jasmussen** [comment]
> Before I review though, how does (if at all) fit in with: #5360?

Ah yes, excellent memory, thanks for bringing that back on my mind, I'd forgotten it totally. 

Re-reading that post, it seems like this PR could still be complementary. I.e. we want to allow themes to load an _editor style_ into the editor, and style every block. If a theme then opts-out of "opinionated styles" as Matías calls it, those styles would simply not load in the editor, and the theme would not have to deal with them, either in the editor or on the frontend. In that vein it would be similar to a theme that has not opted into wide images — they can still style the editor with an editor style.

**Comment by jasmussen** [comment]
Responding separately due to the feedback from two angles:

> I'd tend to think Gutenberg should preserve the accessibility of its own interface, maybe reserving some light background around its controls.

and

> I also wonder a little from an accessibility view but right now we have that with editor styles. Maybe we can consider what boundaries we do/don't let people style?

Both good observations and thoughts. Given the overall positive reception this PR has engendered, I think this is what I'll tackle next. I feel like we have many options on the table, such as just always showing the filled-out buttons behind the side UI, or looking at things like inversion `filter`s or `mix-blend-mode`, to guarantee sufficient contrast in the case of selection styles or placeholder colorings. Those are vague concepts I haven't yet dived into, but they are tools in our toolbelt that we can explore, and I think they can help us get to a point where _the UI_ will always have contrast (for example a total inversion, when necessary, should always have sufficient contrast by definition). 

That said, a badly designed theme will always be able to work around these for non-UI aspects, such as providing light gray text on a white background. Enforcing that seems like a task for the theme-review team, if at all.

**Comment by weavertheme** [comment]
I created a more or less duplicate issue  #6451 (sorry to dup - I find it very hard to search existing issues)

I thnk some of the default colors on these edit controls certainly can use improvement, but I also now think this could be a theme requirement. It is possible to get very nice looking controls using just code listed in #6451. Letting the theme handle control styling alleviates trying to get a one-size-suites all solution.

**Comment by jasmussen** [comment]
Okay, pushed some polish, and I think this is ready for final review.

This PR now:

- Adjusts all the UI on the canvas to be friendly to theme styles. Previously white backgrounds are now transparent, and previously gray buttons are now semi transparent. 
- Multi selections now use mix-blend-mode to ensure that the selection is always visible.

Please review any contrast issues that might have come up as part of this — I've tried my best to match the new opacity based grays to the existing grays, but if there are any regressions, they are easy to adjust.

Default — this should ideally look the same as before:

default

Light theme:

light theme

I also included colors that work based on dark themes. But this is not yet activated anywhere. Essentially we need to apply an `is-dark-theme` class to the `body` element, to invert the UI colors to use shades of white instead of shades of dark gray for the side UI and borders. This PR is ready for that class to exist, but how that class is best added should probably be explored separately. What do you think? Here's how the dark theme looks:

dark theme

Let me know your feedback.

File: editor/components/post-title/style.scss [hunk 1]
```diff
@@ -19,14 +20,19 @@
 
 	&.is-selected .editor-post-title__input,
 	.editor-post-title__input:hover {
-		border: 1px solid $light-gray-500;
+		// use opacity to work in various editor styles
+		border: 1px solid $dark-opacity-light-500;
+
+		.is-dark-theme & {
+			border: 1px solid $light-opacity-light-500;
```
Comment (paulwilde) [inline editor/components/post-title/style.scss]: `border-color: $light-opacity-light-500;` is cleaner.

File: editor/components/block-toolbar/style.scss [hunk 2]
```diff
@@ -24,5 +24,8 @@
 
 .editor-block-toolbar .editor-block-switcher {
 	display: inline-flex;
-	border-left: 1px solid $light-gray-500;
-}
+
+	.edit-post-header-toolbar__block-toolbar & {
+		border-left: 1px solid $light-gray-500;
```
Comment (paulwilde) [inline editor/components/block-toolbar/style.scss]: `border-left-color: $light-gray-500;` is cleaner.

File: editor/components/block-list/style.scss [hunk 3]
```diff
@@ -652,10 +680,18 @@
...
 	width: 100%;
 	background: $white;
 
+	// use opacity to work in various editor styles
+	background-clip: padding-box;
+	box-sizing: padding-box;
+	border: 1px solid $dark-opacity-light-500; 
+
+	.is-dark-theme & {
+		border: 1px solid $light-opacity-light-500;
```
Comment (paulwilde) [inline editor/components/block-list/style.scss]: `border-color: $light-opacity-light-500;` is cleaner.

File: editor/components/block-list/style.scss [hunk 4]
```diff
@@ -250,7 +278,12 @@
 	}
 
 	&.is-shared > .editor-block-list__block-edit:before {
-		outline: 1px dashed $light-gray-500;
+		// use opacity to work in various editor styles
+		outline: 1px dashed $dark-opacity-light-500;
+
+		.is-dark-theme & {
+			outline: 1px dashed $light-opacity-light-500;
```
Comment (paulwilde) [inline editor/components/block-list/style.scss]: `outline-color: $light-opacity-light-500;` is cleaner.

File: editor/components/block-list/style.scss [hunk 5]
```diff
@@ -168,16 +174,26 @@
...
-	&.is-selected > .editor-block-mover:before,
 	&.is-hovered > .editor-block-mover:before {
-		border-right: 1px solid $light-gray-500;
 		right: 0;
+
+		// use opacity to work in various editor styles
+		border-right: 1px solid $dark-opacity-light-500;
+
+		.is-dark-theme & {
+			border-right: 1px solid $light-opacity-light-500;
```
Comment (paulwilde) [inline editor/components/block-list/style.scss]: `border-right-color: $light-opacity-light-500;`.

File: editor/components/block-list/style.scss [hunk 6]
```diff
@@ -168,16 +174,26 @@
...
-	&.is-selected > .editor-block-settings-menu:before,
 	&.is-hovered > .editor-block-settings-menu:before {
-		border-left: 1px solid $light-gray-500;
 		left: 0;
+
+		// use opacity to work in various editor styles
+		border-left: 1px solid $dark-opacity-light-500; 
+
+		.is-dark-theme & {
+			border-left: 1px solid $light-opacity-light-500;
```
Comment (paulwilde) [inline editor/components/block-list/style.scss]: `border-left-color: $light-opacity-light-500;`

File: editor/components/block-list/style.scss [hunk 7]
```diff
@@ -213,7 +229,12 @@
 
 	// focused block-style
 	&.is-selected > .editor-block-list__block-edit:before {
-		outline: 1px solid $light-gray-500;
+		// use opacity to work in various editor styles
+		outline: 1px solid $dark-opacity-light-500;
+	
+		.is-dark-theme & {
+			outline: 1px solid $light-opacity-light-500;
```
Comment (paulwilde) [inline editor/components/block-list/style.scss]: `outline-color: $light-opacity-light-500;`

**Review by karmatosed — APPROVED** [review]

File: edit-post/components/visual-editor/style.scss [hunk 8]
```diff
@@ -67,7 +67,7 @@
 
 // This is a focus style shown for blocks that need an indicator even when in an isEditing state
 // like for example an image block that receives arrowkey focus.
-.edit-post-visual-editor .editor-block-list__block:not( .is-selected )   {
+.edit-post-visual-editor .editor-block-list__block:not( .is-selected ):not( [data-type="core/paragraph"] )   {
```
Comment (noisysocks) [inline edit-post/components/visual-editor/style.scss]: Why not paragraphs?

File: editor/components/block-list/style.scss [hunk 9]
```diff
@@ -1,6 +1,6 @@
 .components-draggable__clone {
 	& > .editor-block-list__block > .editor-block-list__block-draggable {
-		background: $white;
+		background: $white; // @todo: ensure this works with themes that invert the color
```
Comment (noisysocks) [inline editor/components/block-list/style.scss]: Has this been done?

File: editor/components/block-toolbar/style.scss [hunk 10]
```diff
@@ -24,5 +24,8 @@
 
 .editor-block-toolbar .editor-block-switcher {
 	display: inline-flex;
-	border-left: 1px solid $light-gray-500;
-}
+
+	.edit-post-header-toolbar__block-toolbar & {
+		border-left-color: 1px solid $light-gray-500;
+	}
+}
```
Comment (noisysocks) [inline editor/components/block-toolbar/style.scss]: Missing a new line at the end of the file here.

**Review by noisysocks — CHANGES_REQUESTED** [review]
Code LGTM! 🌈✨ I tested this locally and the only bug I noticed was that, when using a dark background, the controls in a column block don't have visible icons:

column buttons

File: edit-post/components/visual-editor/style.scss (hunk 8 above)
Comment (jasmussen) [reply edit-post/components/visual-editor/style.scss]: Good question. This branch has been around long enough that I don't recall adding this — could it have come in as part of a rebase and later been removed? 

In any case I removed it again becaues I couldn't see any regressions as a result.

**Review by noisysocks — APPROVED** [review]

**Comment by jasmussen** [comment]
CC: @jorgefilipecosta — can you rebase your nested blocks for cover image branch and see if this improves things?

**Comment by mcsf** [comment]
I believe this introduced regression #6836, fixed in #6844.

File: edit-post/assets/stylesheets/_z-index.scss [hunk 11]
```diff
@@ -3,8 +3,8 @@
 // value is designed to work with).
 
 $z-layers: (
+	'.editor-block-list__block-edit:before': 0,
```
Comment (youknowriad) [inline edit-post/assets/stylesheets/_z-index.scss]: Not certain why this change was required but it broke the PlainText inside blocks (code block for instance). It should be fixed by this though #6844

**Comment by jasmussen** [comment]
My apologies for the regression, and thank you for fixing it!
